contracts/competition/Competition.sol

```diff
@@ -0,0 +1,169 @@
+import "../access/Governable.sol";
+
+contract Competition is Governable
+{
+    struct Team {
+        address leader;
+        string name;
+        bytes32 referralCode;
+        address[] members;
+    }
+
+    struct Competition {
+        uint start;
+        uint end;
+        uint maxTeamSize;
+        mapping(address => Team) teams;
+        mapping(string => bool) teamNames;
+        mapping(address => address) memberTeams;
+        mapping(address => address) joinRequests;
+    }
+
+    uint public nextCompetitionIndex = 0;
+    mapping(uint => Competition) public competitions;
+    IReferralStorage public referralStorage;
+
+    event TeamCreated(uint index, address leader, string name, bytes32 referral);
+    event JoinRequestCreated(uint index, address member, address leader);
+    event JoinRequestCanceled(uint index, address member);
+    event JoinRequestApproved(uint index, address member, address leader);
+    event MemberRemoved(uint index, address leader, address member);
+    event CompetitionCreated(uint index, uint start, uint end, uint maxTeamSize);
+    event CompetitionUpdated(uint index, uint start, uint end, uint maxTeamSize);
+
+    modifier registrationIsOpen(uint competitionIndex) {
+        require(competitions[competitionIndex].start > block.timestamp, "Registration is closed.");
+        _;
+    }
+
+    modifier isNotMember(uint competitionIndex) {
+        require(competitions[competitionIndex].memberTeams[msg.sender] == address(0), "Team members are not allowed.");
+        _;
+    }
+
+    modifier competitionExists(uint index) {
+        require(competitions[index].start > 0, "The competition does not exist.");
+        _;
+    }
+
+    constructor(IReferralStorage _referralStorage) public {
+        referralStorage = _referralStorage;
+    }
+
+    function createCompetition(uint start, uint end, uint maxTeamSize) external onlyGov {
+        require(start > block.timestamp, "Start time must be in the future.");
+        require(end > start, "End time must be greater than start time.");
+
+        competitions[nextCompetitionIndex] = Competition(start, end, maxTeamSize);
+
+        emit CompetitionCreated(nextCompetitionIndex, start, end, maxTeamSize);
+
+        nextCompetitionIndex++;
+    }
+
+    function updateCompetition(uint index, uint start, uint end, uint maxTeamSize) external onlyGov competitionExists(index) {
+        competitions[index].start = start;
+        competitions[index].end = end;
+        competitions[index].maxTeamSize = maxTeamSize;
+
+        emit CompetitionUpdated(index, start, end, maxTeamSize);
+    }
+
+    function createTeam(uint competitionIndex, string calldata name, bytes32 referralCode) external registrationIsOpen(competitionIndex) isNotMember(competitionIndex) {
+        Competition storage competition = competitions[competitionIndex];
+
+        require(referralStorage.codeOwners(referralCode) != address(0), "Referral code does not exist.");
```

Comment:
@xdev10 you've said we can skip the referral code, do you mean referral code will be optional?